tests/random_attack.py

```diff
@@ -707,6 +739,23 @@ def add_range_column(self, name, rangeobj):
         self.dedup_names()
         self.df.cbind(dt.Frame(rangeobj, names=[name]))
 
+    def set_key_columns(self, keys, names):
+        try:
+            self.df.key = names
+        except ValueError as e:
+            assert str(e) == "Cannot set a key: the values are not unique"
+            return
+
+        nonkeys = sorted(list(set(range(len(self.data))) - set(keys)))
```

Comment:
```suggestion
        nonkeys = set(range(self.ncols)) - set(keys)
        new_col_order = list(keys) + sorted(nonkeys)
```

Reply:
`self.ncols` is calculated on a frame as `self.df.ncols`, while here we're dealing with python data, so it would be nice to stay within the python data context and keep using `len(self.data)` or something like that. The rest, yes, could be simplified.

Reply:
You may adjust the calculation of `.ncols` property, but the code clarity is generally improved when you use `self.ncols` instead of `self.shape[1]` or `len(self.data)`.

Reply:
Yes, I agree, but for some reason `self.ncols` and `self.nrows` are functions of a frame, though `self.types`, `self.names` and `self.data` are functions of python data. 

I propose we keep all datatable stuff under `self.df`, i.e. whenever we do some operations on the frame we use `self.df.ncols` and `self.df.nrows`. When we do operations on the python data along with `self.names`, `self.types`, we also use `self.ncols` defined as `len(self.data)` and `self.nrows` defined as `len(self.data[0])`.